maildir_test.go

```diff
@@ -330,6 +331,85 @@ func TestIllegal(t *testing.T) {
 	}
 }
 
+func TestFolderWithSquareBrackets(t *testing.T) {
+	t.Parallel()
+	root := t.TempDir()
+	name := "[Google Mail].All Mail"
+
+	dir := Dir(filepath.Join(root, name))
+	if err := dir.Init(); err != nil {
+		t.Fatal(err)
+	}
+
+	key := func() string {
+		key, writer, err := dir.Create([]Flag{FlagPassed, FlagReplied})
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer writer.Close()
+		_, err = writer.Write([]byte("this is a message"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		return key
+	}()
+
+	filename, err := dir.Filename(key)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if filename == "" {
+		t.Error("filename should not be empty")
+	}
+}
+
+func TestDifferentSizesOfReaddirChunks(t *testing.T) {
+	totalFiles := 3
+	// don't run this test in // as it modifies a package variable
+	source := t.TempDir()
+
+	dir := Dir(source)
+	if err := dir.Init(); err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < totalFiles; i++ {
+		makeDelivery(t, dir, fmt.Sprintf("here is message number %d", i))
+	}
+
+	// grab keys
+	keys, err := dir.Unseen()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// avoid the filename guesser
+	for _, key := range keys {
+		err := dir.SetFlags(key, []Flag{FlagPassed, FlagReplied})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	previousReaddirChunk := readdirChunk
+
+	// try different sizes of chunks
+	for chunkSize := 0; chunkSize <= totalFiles+1; chunkSize++ {
+		readdirChunk = chunkSize
+		for _, key := range keys {
+			filename, err := dir.Filename(key)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if filename == "" {
+				t.Errorf("cannot find filename for key %q", key)
+			}
+		}
+	}
+	// set it back to normal for the following tests
+	readdirChunk = previousReaddirChunk
```

Comment:
Maybe put this in a `defer` block, so that we restore the old value even on `t.Fatal`?